### PR TITLE
ui: Redirect to latest run from repository page

### PR DESCRIPTION
**Pull request by mmurto**
This is waiting for #1469.

See individual commits.

image

image

image

### CODE REVIEW

**Comment by sschuberth** [comment]
Only some nits from looking at the screenshots:
- The menu item is called "All Runs", but at least in the second screenshots the title displayed for that page is just "Runs". We should align this, and I have a tendency towards calling the menu item also just "Runs" for simplicity (unless you want to emphasize the "All" for some reason, i.e. if there also would be not-*all*-runs view).
- I like the idea to consolidate editing and deleting in a new menu entry. However, as this is not only about editing, but also deleting, maybe call the menu entry "Manage" or so instead?

**Comment by mmurto** [comment]
> Only some nits from looking at the screenshots:
> 
> * The menu item is called "All Runs", but at least in the second screenshots the title displayed for that page is just "Runs". We should align this, and I have a tendency towards calling the menu item also just "Runs" for simplicity (unless you want to emphasize the "All" for some reason, i.e. if there also would be not-_all_-runs view).

I agree that it should be aligned for the repository view, so in the second screenshot the sidebar item should be "Runs" like the heading is. In the run view, so in the first screenshot, I think "All Runs" makes more sense it's a link from one run to all runs.

> * I like the idea to consolidate editing and deleting in a new menu entry. However, as this is not only about editing, but also deleting, maybe call the menu entry "Manage" or so instead?

Agree. "Settings" feels natural to me, I'll try that.

**Comment by mnonnenmacher** [comment]
I very often want to see the list of runs and not the results of the last run, would it be an alternative to just have separate buttons in the repository list to either got to the list or go to the last run?
Also, having the "All Runs" button below the "Overview" breaks the hierarchy to me: The list of runs is still a logical parent of the run details, so the button should be on top and maybe separated with a divider.

**Comment by sschuberth** [comment]
> I very often want to see the list of runs and not the results of the last run

Interesting, we've received pretty clear feedback from our users that they want / expected the opposite. Can you explain *why* you want to see the list of all runs, what is the information you're looking for? It it about the success / failure rate, or how the last run compares to the previous runs in that regard?

Depending on your answer, maybe we could tweaks things to not unconditionally show the last run, but the last *successful* run? Thought that logic could be more confusing to users.

**Comment by sschuberth** [comment]
> The list of runs is still a logical parent of the run details

I believe this is one of the cases where we should be free to relax hierarchies implied by backend code in the UI for the sake of usability and convenience. Without knowing anything about the code and just looking at the UI, I think having an "Overview" first and then next "All Runs" to drill into more details is just fine.

**Comment by sschuberth** [comment]
@mmurto how about renaming the "Settings" menu from the bottom-left of your first screenshots to "Repository" as follows:

```
Repository
  Secrets
  Users
  Settings
```

This avoids the "Repository" word duplication for the nested entries and makes more clear that no run-related stuff is in there.

**Comment by mmurto** [comment]
> > I very often want to see the list of runs and not the results of the last run
> 
> Interesting, we've received pretty clear feedback from our users that they want / expected the opposite. Can you explain _why_ you want to see the list of all runs, what is the information you're looking for? It it about the success / failure rate, or how the last run compares to the previous runs in that regard?

I think this also depends on who the primary user that ends up on the repository page is. In my view that user is likely to be the most interested in the state of their repository's state, so showing the most recent run prominently makes sense to me. Now I'm mostly guessing (we should add analytics!), but I'd imagine users who are more interested in diving deep into the past runs use the UI more than other users, so they know their way around it more.

> @mmurto how about renaming the "Settings" menu from the bottom-left of your first screenshot to "Repository" as follows:
> 
> ```
> Repository
>   Secrets
>   Users
>   Settings
> ```
> 
> This avoids the "Repository" word duplication for the nested entries and makes more clear that no run-related stuff is in there.
> 
> _Edit:_ And for "Settings" I'd use https://lucide.dev/icons/settings as the icon.

Agree.

**Comment by sschuberth** [comment]
> I'd imagine users who are more interested in diving deep into the past runs use the UI more than other users, so they know their way around it more.

Exactly my thoughts as well. For more technical users / admins it is reasonable to navigate in the UI a bit more.

**Comment by mnonnenmacher** [comment]
>Interesting, we've received pretty clear feedback from our users that they want / expected the opposite. Can you explain why you want to see the list of all runs, what is the information you're looking for? It it about the success / failure rate, or how the last run compares to the previous runs in that regard?

I think because I'm looking at it from an admin perspective, I'm usually more interested to see how stable the results are, how often the repo is scanned, and so on.
But also from a user perspective, as long as you manually trigger all runs the last one maybe stands out, but as soon as you trigger them from CI, e.g., from each PR, the last run becomes arbitrary and you might be more interested in the last run for the main branch.

>Depending on your answer, maybe we could tweaks things to not unconditionally show the last run, but the last successful run? Thought that logic could be more confusing to users.

Yes, I think that would be confusing.

> >The list of runs is still a logical parent of the run details

>I believe this is one of the cases where we should be free to relax hierarchies implied by backend code in the UI for the sake of usability and convenience. Without knowing anything about the code and just looking at the UI, I think having an "Overview" first and then next "All Runs" to drill into more details is just fine.

I didn't mean it from a technical perspective but from how I understand the content. To me a list is a parent object of one item in the list. That's why I would prefer to have the "All runs" link on top of the sidebar, then its position would also not change after clicking on it.
And BTW, like I mentioned in #1039, I would really like if next to the "Overview" button some context information would be shown. This would also justify separating "Overview" from "All runs" with a divider.

>I think this also depends on who the primary user that ends up on the repository page is. In my view that user is likely to be the most interested in the state of their repository's state, so showing the most recent run prominently makes sense to me. Now I'm mostly guessing (we should add analytics!), but I'd imagine users who are more interested in diving deep into the past runs use the UI more than other users, so they know their way around it more.

Still, what about my proposal to offer both options? For example, the number of runs ("3" in the image below) could be clickable and lead directly to the history:
image

**Comment by mmurto** [comment]
> > Interesting, we've received pretty clear feedback from our users that they want / expected the opposite. Can you explain why you want to see the list of all runs, what is the information you're looking for? It it about the success / failure rate, or how the last run compares to the previous runs in that regard?
> 
> I think because I'm looking at it from an admin perspective, I'm usually more interested to see how stable the results are, how often the repo is scanned, and so on. But also from a user perspective, as long as you manually trigger all runs the last one maybe stands out, but as soon as you trigger them from CI, e.g., from each PR, the last run becomes arbitrary and you might be more interested in the last run for the main branch.

One idea I've been toying with is that it could be nice to build the UI (with proper endpoints for it) not around ORT runs, but commits that have been analyzed. Inside commits you could then have multiple runs and the commits could also be grouped by the branch they belong to.

> Still, what about my proposal to offer both options? For example, the number of runs ("3" in the image below) could be clickable and lead directly to the history: image

Good idea, I added that! I originally tried some Actions buttons like in other tables, but the layout is starting to be to narrow for it, so we'll need to widen it.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/runs/index.tsx [hunk 1]
```diff
@@ -47,11 +47,11 @@ import {
 } from '@/components/ui/tooltip';
 import { toast } from '@/lib/toast';
 import { paginationSearchParameterSchema } from '@/schemas';
-import { RepositoryRunsTable } from '../-components/repository-runs-table';
+import { RepositoryRunsTable } from '../../-components/repository-runs-table';
```
Comment (Etsija) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/runs/index.tsx:45]: Should this now actually be an absolute link `@/...` as it's not anymore "close" to the page where it's used?

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+      useRepositoriesServiceGetRepositoryById,
+      params.orgId,
+      params.productId,
+      params.repoId,
+    ],
+    queryFn: async () =>
+      await RepositoriesService.getRepositoryById({
+        repositoryId: Number.parseInt(params.repoId),
+      }),
+  });
```
Comment (Etsija) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx:82]: Is there a particular reason why a suspense hook is not used here? Also, can you explain why there are extra elements in the query key, which are not parameters of the query?

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    await context.queryClient.ensureQueryData({
+      queryKey: [useRepositoriesServiceGetRepositoryByIdKey, params.repoId],
+      queryFn: () =>
+        RepositoriesService.getRepositoryById({
+          repositoryId: Number.parseInt(params.repoId),
+        }),
+    });
+  },
+  component: RepositorySettingsPage,
+});
```
Comment (Etsija) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx:275]: Same here: why not use the prefetch hook? Query key is also now different from the one above.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx (hunk 2 above)
Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx:82]: No particular reason, as the settings page is just renamed [edit page](https://github.com/eclipse-apoapsis/ort-server/blame/main/ui/src/routes/_layout/organizations/%24orgId/products/%24productId/repositories/%24repoId/_repo-layout/edit.tsx) with the addition of the danger zone including a delete button.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx (hunk 3 above)
Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/settings.tsx:275]: True, seems to be a bug in the old edit page, which got carried over here. I'll add a fixing comment.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/runs/index.tsx (hunk 1 above)
Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/runs/index.tsx:45]: > Should this now actually be an absolute import `@/...` as it's not anymore "close" to the page where it's used?

I think it's still relatively close, as the absolute import would be `@/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/repository-runs-table`. What do you think?

Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_repo-layout/runs/index.tsx:45]: Well, yeah, probably better keep it as-is.

**Review by Etsija — APPROVED** [review]

**Pull request merged**